[GStreamer][MediaStream] Leverage GStreamerCapturer on mock sources

philn opened the pull request:
#### 79141ab69a81e754ecc39b8f4f7ab87279401404
<pre>
[GStreamer][MediaStream] Leverage GStreamerCapturer on mock sources
<a href="https://bugs.webkit.org/show_bug.cgi?id=244003">https://bugs.webkit.org/show_bug.cgi?id=244003</a>

Reviewed by Xabier Rodriguez-Calvar.

The GStreamer mock audio and video sources now feed a capturer pipeline, in order to have a similar
behaviour to the non-mock sources. In order to make this possible, we now have a new
GstDeviceProvider that creates a GstDevice for each mock device of the mock realtime media source
center. Each mock device is associated with an appsrc element, to which mock GstSamples are pushed.

* LayoutTests/fast/mediastream/device-change-event-2.html:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr&lt;GstDevice&gt;):
(WTF::derefGPtr&lt;GstDevice&gt;):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::captureDevicesChanged):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
(WebCore::GStreamerAudioCaptureSource::GStreamerAudioCaptureSource):
(WebCore::GStreamerAudioCaptureSource::startProducingData):
(WebCore::GStreamerAudioCaptureSource::stopProducingData):
(WebCore::GStreamerAudioCaptureSource::newSampleCallback): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.cpp:
(WebCore::GStreamerAudioCapturer::GStreamerAudioCapturer):
(WebCore::GStreamerAudioCapturer::~GStreamerAudioCapturer):
(WebCore::GStreamerAudioCapturer::setSinkAudioCallback):
(WebCore::GStreamerAudioCapturer::createConverter):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDevice.h:
(WebCore::GStreamerCaptureDevice::GStreamerCaptureDevice):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::GStreamerCaptureDeviceManager):
(WebCore::GStreamerCaptureDeviceManager::~GStreamerCaptureDeviceManager):
(WebCore::GStreamerCaptureDeviceManager::stopMonitor):
(WebCore::GStreamerCaptureDeviceManager::devicesChanged):
(WebCore::GStreamerCaptureDeviceManager::gstreamerDeviceWithUID):
(WebCore::GStreamerCaptureDeviceManager::captureDevices):
(WebCore::GStreamerCaptureDeviceManager::addDevice):
(WebCore::GStreamerCaptureDeviceManager::refreshCaptureDevices):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.cpp:
(WebCore::GStreamerCapturer::GStreamerCapturer):
(WebCore::GStreamerCapturer::createSource):
(WebCore::GStreamerCapturer::caps):
(WebCore::GStreamerCapturer::setupPipeline):
(WebCore::GStreamerCapturer::start):
(WebCore::GStreamerCapturer::stop):
(WebCore::GStreamerCapturer::play): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.cpp: Added.
(webkitMockDeviceCreateElement):
(webkit_mock_device_class_init):
(webkitMockDeviceCreate):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.h: Added.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDeviceProvider.cpp: Added.
(webkitMockDeviceProviderProbe):
(webkit_mock_device_provider_class_init):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDeviceProvider.h: Added.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.cpp:
(WebCore::GStreamerVideoCaptureSource::~GStreamerVideoCaptureSource):
(WebCore::GStreamerVideoCaptureSource::startProducingData):
(WebCore::GStreamerVideoCaptureSource::processNewFrame): Deleted.
(WebCore::GStreamerVideoCaptureSource::newSampleCallback): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.h:
(WebCore::GStreamerVideoCaptureSource::capturer):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
(WebCore::GStreamerVideoCapturer::GStreamerVideoCapturer):
(WebCore::GStreamerVideoCapturer::~GStreamerVideoCapturer):
(WebCore::GStreamerVideoCapturer::setSinkVideoFrameCallback):
(WebCore::GStreamerVideoCapturer::getBestFormat):
(WebCore::GStreamerVideoCapturer::reconfigure):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeAudioSourceGStreamer.cpp:
(WebCore::MockRealtimeAudioSourceGStreamer::startProducingData):
(WebCore::MockRealtimeAudioSourceGStreamer::stopProducingData):
(WebCore::MockRealtimeAudioSourceGStreamer::render):
(WebCore::MockRealtimeAudioSourceGStreamer::reconfigure):
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeVideoSourceGStreamer.cpp:
(WebCore::MockRealtimeVideoSourceGStreamer::startProducingData):
(WebCore::MockRealtimeVideoSourceGStreamer::stopProducingData):
(WebCore::MockRealtimeVideoSourceGStreamer::updateSampleBuffer):
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeVideoSourceGStreamer.h:
* Source/WebCore/platform/mock/MockRealtimeAudioSource.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockRealtimeMediaSourceCenter::displayDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:

Canonical link: <a href="https://commits.webkit.org/265250@main">https://commits.webkit.org/265250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f579cf8d4cf7e70ab3b04da0ed651f85449b8963

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10349 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10583 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10854 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9943 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10361 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12655 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12908 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11290 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8721 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12389 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/10454 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8593 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9375 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16644 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9687 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12774 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9965 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9270 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2490 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13375 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9816 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
```diff
@@ -87,7 +87,7 @@ AudioCaptureFactory& GStreamerAudioCaptureSource::factory()
 GStreamerAudioCaptureSource::GStreamerAudioCaptureSource(GStreamerCaptureDevice device, MediaDeviceHashSalts&& hashSalts)
     : RealtimeMediaSource(device, WTFMove(hashSalts))
-    , m_capturer(makeUnique<GStreamerAudioCapturer>(device))
+    , m_capturer(makeUnique<GStreamerAudioCapturer>(WTFMove(device)))
```
Did you by any change forget to add the `&&` do the device parameter?

philn replied:
Yes

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDevice.h:
```diff
@@ -26,15 +26,13 @@
     GStreamerCaptureDevice(GRefPtr<GstDevice>&& device, const String& persistentId, DeviceType type, const String& label, const String& groupId = emptyString())
         : CaptureDevice(persistentId, type, label, groupId)
-        , m_device(device)
+        , m_device(WTFMove(device))
```
:100:

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
```diff
@@ -191,7 +224,8 @@ void GStreamerCaptureDeviceManager::refreshCaptureDevices()
     while (devices) {
-        addDevice(GST_DEVICE_CAST(devices->data));
+        auto device = adoptGRef(GST_DEVICE_CAST(devices->data));
+        addDevice(WTFMove(device));
```
you can just do `addDevice(adoptGRef(GST_DEVICE_CAST(devices->data)))`

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
```diff
@@ -27,22 +27,28 @@
     const Vector<CaptureDevice>& captureDevices() final;
     virtual CaptureDevice::DeviceType deviceType() = 0;
 
+    // RealtimeMediaSourceCenter::Observer interface
```
Missing period.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.cpp:
```diff
@@ -0,0 +1,88 @@
+#include "GStreamerMockDevice.h"
+
+#include "CaptureDevice.h"
+#include "GRefPtrGStreamer.h"
```
This is included in common, right?

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.cpp:
```diff
@@ -0,0 +1,88 @@
+#include "CaptureDevice.h"
+#include "GRefPtrGStreamer.h"
+#include "GStreamerCommon.h"
+#include "GUniquePtrGStreamer.h"
```
Ditto

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.cpp:
```diff
@@ -0,0 +1,88 @@
+    GUniquePtr<GstStructure> properties(gst_structure_new("webkit-mock-device", "persistent-id", G_TYPE_STRING, captureDevice.persistentId().ascii().data(), "is-default", G_TYPE_BOOLEAN, captureDevice.isDefault(), nullptr));
+    auto* device = WEBKIT_MOCK_DEVICE_CAST(g_object_new(GST_TYPE_MOCK_DEVICE, "display-name", displayName.ascii().data(), "device-class", deviceClass, "caps", caps.get(), "properties", properties.get(), nullptr));
+    gst_object_ref_sink(device);
+    return GST_DEVICE_CAST(device);
```
I am not objecting to this but if I were you I would have created a smart ptr and then leaked the ref when returning to avoid the ref sink

philn replied:
Isn't that a bit convoluted?

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.cpp:
```diff
@@ -0,0 +1,88 @@
+};
+
+GST_DEBUG_CATEGORY_STATIC(webkitGstMockDeviceDebug);
+#define GST_CAT_DEFAULT webkitGstMockDeviceDebug
```
Where is this initialized?

philn replied:
In the `WEBKIT_DEFINE_TYPE_WITH_CODE` macro expansion.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDeviceProvider.cpp:
```diff
@@ -0,0 +1,73 @@
+};
+
+GST_DEBUG_CATEGORY_STATIC(webkitGstMockDeviceProviderDebug);
+#define GST_CAT_DEFAULT webkitGstMockDeviceProviderDebug
```
Where's this initialized?

philn replied:
In the `WEBKIT_DEFINE_TYPE_WITH_CODE` macro expansion.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDeviceProvider.cpp:
```diff
@@ -0,0 +1,73 @@
+        devices = g_list_prepend(devices, webkitMockDeviceCreate(device));
+    for (auto& device : sourceCenter.displayDevices())
+        devices = g_list_prepend(devices, webkitMockDeviceCreate(device));
+    devices = g_list_reverse(devices);
```
Does order matter, traverse the list backwards or remove elements? We could even avoid reversing the list or even turning it into a GSList and saving the prev pointer.

philn replied:
I used the recommended approach from the GList docs, GSList not an option I think, since the probe function should return a GList.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.cpp:
```diff
@@ -117,7 +117,7 @@ GStreamerVideoCaptureSource::GStreamerVideoCaptureSource(String&& deviceID, Atom
 GStreamerVideoCaptureSource::GStreamerVideoCaptureSource(GStreamerCaptureDevice device, MediaDeviceHashSalts&& hashSalts)
     : RealtimeVideoCaptureSource(device, WTFMove(hashSalts), { })
-    , m_capturer(makeUnique<GStreamerVideoCapturer>(device))
+    , m_capturer(makeUnique<GStreamerVideoCapturer>(WTFMove(device)))
```
As I asked before, did you maybe forget to add the && to the device parameter?

philn replied:
Yes

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
```diff
@@ -52,6 +55,30 @@ GStreamerVideoCapturer::GStreamerVideoCapturer(const char* sourceFactory, Captur
+        capturer->m_sinkVideoFrameCallback.second(VideoFrameGStreamer::create(WTFMove(gstSample), WebCore::FloatSize(), presentationTime));
+        return GST_FLOW_OK;
+    }), this);
+    m_sinkVideoFrameCallback.first = handler;
```
Why don't you just assign the result of the function to first, avoiding the handler var?

philn replied:
No reason :)

philn committed:
[GStreamer][MediaStream] Leverage GStreamerCapturer on mock sources
https://bugs.webkit.org/show_bug.cgi?id=244003

Reviewed by Xabier Rodriguez-Calvar.

The GStreamer mock audio and video sources now feed a capturer pipeline, in order to have a similar
behaviour to the non-mock sources. In order to make this possible, we now have a new
GstDeviceProvider that creates a GstDevice for each mock device of the mock realtime media source
center. Each mock device is associated with an appsrc element, to which mock GstSamples are pushed.

* LayoutTests/fast/mediastream/device-change-event-2.html:
* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr<GstDevice>):
(WTF::derefGPtr<GstDevice>):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::captureDevicesChanged):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
(WebCore::GStreamerAudioCaptureSource::GStreamerAudioCaptureSource):
(WebCore::GStreamerAudioCaptureSource::startProducingData):
(WebCore::GStreamerAudioCaptureSource::stopProducingData):
(WebCore::GStreamerAudioCaptureSource::newSampleCallback): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.cpp:
(WebCore::GStreamerAudioCapturer::GStreamerAudioCapturer):
(WebCore::GStreamerAudioCapturer::~GStreamerAudioCapturer):
(WebCore::GStreamerAudioCapturer::setSinkAudioCallback):
(WebCore::GStreamerAudioCapturer::createConverter):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDevice.h:
(WebCore::GStreamerCaptureDevice::GStreamerCaptureDevice):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::GStreamerCaptureDeviceManager):
(WebCore::GStreamerCaptureDeviceManager::~GStreamerCaptureDeviceManager):
(WebCore::GStreamerCaptureDeviceManager::stopMonitor):
(WebCore::GStreamerCaptureDeviceManager::devicesChanged):
(WebCore::GStreamerCaptureDeviceManager::gstreamerDeviceWithUID):
(WebCore::GStreamerCaptureDeviceManager::captureDevices):
(WebCore::GStreamerCaptureDeviceManager::addDevice):
(WebCore::GStreamerCaptureDeviceManager::refreshCaptureDevices):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.cpp:
(WebCore::GStreamerCapturer::GStreamerCapturer):
(WebCore::GStreamerCapturer::createSource):
(WebCore::GStreamerCapturer::caps):
(WebCore::GStreamerCapturer::setupPipeline):
(WebCore::GStreamerCapturer::start):
(WebCore::GStreamerCapturer::stop):
(WebCore::GStreamerCapturer::play): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.cpp: Added.
(webkitMockDeviceCreateElement):
(webkit_mock_device_class_init):
(webkitMockDeviceCreate):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDevice.h: Added.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDeviceProvider.cpp: Added.
(webkitMockDeviceProviderProbe):
(webkit_mock_device_provider_class_init):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMockDeviceProvider.h: Added.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.cpp:
(WebCore::GStreamerVideoCaptureSource::~GStreamerVideoCaptureSource):
(WebCore::GStreamerVideoCaptureSource::startProducingData):
(WebCore::GStreamerVideoCaptureSource::processNewFrame): Deleted.
(WebCore::GStreamerVideoCaptureSource::newSampleCallback): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.h:
(WebCore::GStreamerVideoCaptureSource::capturer):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
(WebCore::GStreamerVideoCapturer::GStreamerVideoCapturer):
(WebCore::GStreamerVideoCapturer::~GStreamerVideoCapturer):
(WebCore::GStreamerVideoCapturer::setSinkVideoFrameCallback):
(WebCore::GStreamerVideoCapturer::getBestFormat):
(WebCore::GStreamerVideoCapturer::reconfigure):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeAudioSourceGStreamer.cpp:
(WebCore::MockRealtimeAudioSourceGStreamer::startProducingData):
(WebCore::MockRealtimeAudioSourceGStreamer::stopProducingData):
(WebCore::MockRealtimeAudioSourceGStreamer::render):
(WebCore::MockRealtimeAudioSourceGStreamer::reconfigure):
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeAudioSourceGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeVideoSourceGStreamer.cpp:
(WebCore::MockRealtimeVideoSourceGStreamer::startProducingData):
(WebCore::MockRealtimeVideoSourceGStreamer::stopProducingData):
(WebCore::MockRealtimeVideoSourceGStreamer::updateSampleBuffer):
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeVideoSourceGStreamer.h:
* Source/WebCore/platform/mock/MockRealtimeAudioSource.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockRealtimeMediaSourceCenter::displayDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:

Canonical link: https://commits.webkit.org/265250@main
